### DISCUSSION TITLE
defineCollection: allow frontmatter modifications

### DISCUSSION THREAD

**Discussion by jamesarosen** [withastro/roadmap#514] [Proposal]
## Summary

I'd like to be able to perform modifications to the data via `defineCollection`

## Background & Motivation

When migrating from other blogging platforms (e.g. Jekyll), the Markdown frontmatter may have a different structure even though it represents the same concepts. It would be nice to be able to use these files directly and apply any necessary conversions centrally.

## Goals

* make it easier to migrate content from other blogging platforms

## Example

This example converts Jekyll-compatible frontmatter to Astro-compatible. Notably, Jekyll uses a `permalink` that can have leading and trailing slashes, whereas Astro uses a `slug` that has neither.

```ts
const blogCollection = defineCollection({
	schema: z
		.object({
			permalink: z.string(),
			title: z.string(),
		})
		.transform((frontmatter) => {
			const result = {
				...frontmatter,
				// convert from permalink to slug:
				slug: frontmatter.permalink.replace(/^\/?(.+?)\/?$/, '$1'),
			}
			return result
		}),
})
```

## References

* https://github.com/una/una.im/pull/2 -- an example of such a conversion with the currently-available API
* https://github.com/withastro/roadmap/discussions/489 -- a proposal that uses different concepts and APIs to achieve some similar goals
* https://github.com/withastro/roadmap/discussions/415

**Comment by bholmesdev** [comment]
Hey James! I agree frontmatter modifications are a great feature. I'm a bit confused what is missing in today's APIs to achieve this though? As written, that example you gave should work in content collections already. We could use better documentation though.

**Reply by jamesarosen** [reply]
You're absolutely right. https://stackblitz.com/edit/node-qhux4n shows this off using Astro v2.1. I'm not sure what we were doing wrong when I wrote this.

**Reply by bholmesdev** [reply]
@jamesarosen Haha no problem! I'll let the docs team know in case there's some room for improvement.
